### DISCUSSION TITLE
How to get shared clipboard and good performance?

### DISCUSSION THREAD

**Discussion by LorenzoBettini** [quickemu-project/quickemu#1043] [Quickemu] [answered]
#### Expected behaviour

First of all, I'm only experimenting with Linux VMs. And the spice agent software is installed on the host.

I seem to understand that shared clipboard only works when using `--display spice`. Indeed, it does (only when using an X11, not when the host runs a Wayland session). However, the performance is really bad (3d effects are really slow). Then I tried `--display spice-app`, however, in that case, the VM window flashes and has serious graphical problems that do not even allow me to log into the VM.

Long story short: is there a way to have a shared clipboard in the Linux VM without sacrificing performance?

#### Quickemu output

Run `quickemu` or `quickget` and paste the output here.

```
❯ quickemu --version                                           
4.9.1
```

#### Linux Distribution & Kernel

Run the following and paste the output.

```bash
lsb_release --all

LSB Version:    n/a
Distributor ID: EndeavourOS
Description:    EndeavourOS Linux
Release:        rolling
Codename:       rolling

uname -a

Linux eos-optiplex 6.6.16-1-lts #1 SMP PREEMPT_DYNAMIC Mon, 05 Feb 2024 21:20:21 +0000 x86_64 GNU/Linux
```

**Comment by TuxVinyards** [comment, marked as answer]
You are never going to get fast slick graphical desktop performance.  The Spice feature is useful for pasting the file. It ends up in the Downloads folder.  You can use qqX to set up a shared drive that all the different VM's can access. Also works in SDL mode too ...  That's about it.

https://github.com/HikariKnight has a 3D pass through project you might want to look at.  I haven't tried it.  More aimed at gamers.

**Comment by LorenzoBettini** [comment]
> You are never going to get fast slick graphical desktop performance. 

You mean with spice, right?
With the default `--display sdl` the performance are great.

**Comment by TuxVinyards** [comment]
> "With the default --display sdl the performance are great."

Well, I wouldn't say that the performance is ever going to bowl you over, but yes, I think it is a bit crisper with SDL.  And you can set things up to launch at a given screen size too ...

**Comment by LorenzoBettini** [comment]
I wouldn't call it just a bit crispier ;)
On my computers, it doesn't even feel like a virtual machine: it's really smooth. And only with an Intel HD Graphics.
When using the spice display instead things degrade a lot.

**Comment by TuxVinyards** [comment]
Good to hear :+1:

**Comment by muhmud** [comment]
I have the same issue, `sdl` works really great, like native performance, but I can't get the clipboard to work, which I really need. Meanwhile, I can copy/paste with `spice`, but the performance is disappointing.

Is there anyway to get shared clipboard to work in `sdl`?

**Comment by TuxVinyards** [comment]
No. That's it at present.  Those are the tools that  we have. But at least we have those ....

Edit & Update to the above and probably over-quick reply:

On the horizon is  SDL3.   This is not yet in general release and is still somewhat a work in progress.  

Things could change in the next few months.   @LorenzoBettini @muhmud  if you are able to keep an eye on this and post up here, maybe, I think this could get implemented.

Currently, I am working to get the existing quickemu and quickget running as it should within qqX.  This is taking up most of my time. But there is quite a bit of discussion going on behind the scenes about how to take the whole project forward.

![SDL3-migration](https://github.com/quickemu-project/quickemu/assets/3956806/71ceafc0-89ca-49c6-a607-30b15ff01574)

We would probably need to see SDL3 generally available before we can do much.  Note in screenshot of Ubuntu 23.10 that only SDL2 is possible. Although, some kind of test for system availability could be worked ...  Thinking aloud a bit here; problems with LTS and 6 month update distros etc

According to libsdl here <https://github.com/libsdl-org/SDL/issues/8338>  a clipboard is making it's way into circulation.

:rocket:

**Comment by thelicato** [comment]
Hello everyone, 
SDL 3 has been officially released (https://github.com/libsdl-org/SDL/releases/tag/release-3.2.2, already reached 3.2.2 due to some bugfixes). Is anyone interested in building a SDL3 display for QEMU?
I have the will, but not so much time and the required skills to do it so I'm here to check is someone wants to work together on it

**Comment by TuxVinyards** [comment]
@thelicato  :wave:  

I am seeing "Allow the window manager to position windows that don't specify a position on Linux" as feature to 3.2.2.  I was pleased to see this out the other day too.  No more having to drag the window out of the bottom right and into the centre, hopefully.

I don't think we are seeing copy and paste as yet.  I may have missed that bit but I have been looking  ....

Something new,  for [qqX](https://github.com/TuxVinyards/qqX/wiki/Disks-and-Resizing#other-tools) users, is that you can now access a shared public drive when running with SDL, which I am finding to be very useful.

> Is anyone interested in building a SDL3 display for QEMU?

I am keeping an eye out on my Arch distros for new developments.  I don't think that the Qemu team will be letting the grass grow under their feet on getting this implemented.

**Reply by thelicato** [reply]
> @thelicato 👋
> 
> I am seeing "Allow the window manager to position windows that don't specify a position on Linux" as feature to 3.2.2. I was pleased to see this out the other day too. No more having to drag the window out of the bottom right and into the centre, hopefully.
> 
> I don't think we are seeing copy and paste as yet. I may have missed that bit but I have been looking ....
> 

The clipboard features have been introduced in ``3.2.0`` (https://github.com/libsdl-org/SDL/releases/tag/release-3.2.0 and https://wiki.libsdl.org/SDL3/CategoryClipboard)

> Something new, for [qqX](https://github.com/TuxVinyards/qqX/wiki/Disks-and-Resizing#other-tools) users, is that you can now access a shared public drive when running with SDL, which I am finding to be very useful.
> 
> > Is anyone interested in building a SDL3 display for QEMU?
> 
> I am keeping an eye out on my Arch distros for new developments. I don't think that the Qemu team will be letting the grass grow under their feet on getting this implemented.

I am afraid they will. I spoke to some developers on the Matrix room and their plan is to focus more on the "backend" features and less on the displays. In fact, since they implemented the ``D-Bus`` display the idea is to let the clients implement their own display.

**Reply by TuxVinyards** [reply]
> The clipboard features have been introduced in 3.2.0 (https://github.com/libsdl-org/SDL/releases/tag/release-3.2.0 and https://wiki.libsdl.org/SDL3/CategoryClipboard)

Woo!

It's on my list.  Just tying up a new qqX release. When that's done I'll have a look and see what I can do.

**Comment by lj3954** [comment]
SDL docs state that clipboard has been available since 2.0.0: https://wiki.libsdl.org/SDL2/SDL_GetClipboardText
3.2.0 does not implement the clipboard, just enable other data types through the API (as opposed to just text).

**Reply by thelicato** [reply]
> SDL docs state that clipboard has been available since 2.0.0: https://wiki.libsdl.org/SDL2/SDL_GetClipboardText 3.2.0 does not implement the clipboard, just enable other data types through the API (as opposed to just text).

Yes, you are right about that. Also the ``ClipboardEvent`` is a new feature: https://wiki.libsdl.org/SDL3/SDL_ClipboardEvent
This feels particularly interesting to implement a good display for QEMU VMs
